### ISSUE TITLE
Admin PirepFiled Notification Fails

### ISSUE THREAD

**Issue by FatihKoz** [phpvms/phpvms#1882]
When a pirep is filed, admins do get mails/notifications but somehow it fails while serializing the SimBriefXML (relationship of a pirep)

```
[2024-09-17 00:17:37] production.EMERGENCY: Error emailing admin (hello@kitty.com.mr). Error=Serialization of 'App\Models\SimBriefXML' is not allowed
```

**Version**
Version 7.0.25-dev.g14cc6ceb55 
_(newer builds should have this problem too as nothing changed in notifications recently)_

**To Reproduce**
1. Enable admin notifications via mail
2. Create a simbrief plan for a flight
3. Send the pirep
4. Check logs

**Expected behavior**
Notifications should be sent to admins as it does not need the SimBrief and/or any other relationships to proceed.

Possible related spots
https://github.com/nabeelio/phpvms/blob/dev/app/Notifications/Messages/PirepFiled.php#L27

https://github.com/nabeelio/phpvms/blob/dev/app/Notifications/NotificationEventsHandler.php#L71

**Comment by ProAviaAZ** [comment]
Dispo was kind enough to open this issue for me. 
Additional info...

phpVMS 7.0.25 from 09/11 updated on 09/15
Previous phpVMS 7.0.12 from 07/17

vmsACARS 1.2.0.2449

Emails stopped being received by admins around 09/11/2024 for vmsACARS pireps.

Submitting a manual pirep with and without using SimBrief results in an email being sent.

```
Manual no SimBrief
[2024-09-17 20:08:32] production.INFO: PIREP Custom Fields  
[2024-09-17 20:08:32] production.INFO: Finance: PIREP: J4l6A9eWrO1PzpMN, flight fares:  [{"fare_id":1,"count":0},{"fare_id":2,"count":0}] 
[2024-09-17 20:08:32] production.INFO: Finance: PIREP: J4l6A9eWrO1PzpMN, fare count: {"fare_id":1,"count":0}  
[2024-09-17 20:08:32] production.INFO: Finance: PIREP: J4l6A9eWrO1PzpMN, fare count: {"fare_id":2,"count":0}  
[2024-09-17 20:08:33] production.DEBUG: Looking for RSK  
[2024-09-17 20:08:33] production.DEBUG: name: RSK - 36.7484x-108.0989  
[2024-09-17 20:08:33] production.INFO: New PIREP filed, pirep_id: J4l6A9eWrO1PzpMN  
[2024-09-17 20:08:33] production.INFO: Bid for user: FA003 on flight FA742  
[2024-09-17 20:08:33] production.INFO: NotificationEvents::onPirepFile: J4l6A9eWrO1PzpMN filed  
[2024-09-17 20:08:33] production.INFO: Disposable Basic | Pirep:J4l6A9eWrO1PzpMN FILED, Change STATE of N738FA to PARKED  
[2024-09-17 20:08:33] production.INFO: Disposable Special | Maintenance Records for N738FA (ID:5) increased Pirep ID:J4l6A9eWrO1PzpMN FILED  

Manual with SimBrief
[2024-09-17 21:28:40] production.INFO: PIREP Custom Fields  
[2024-09-17 21:28:40] production.INFO: Finance: PIREP: RBJeLydD6ebz4VK8, flight fares:  [{"fare_id":1,"count":152},{"fare_id":2,"count":8941}] 
[2024-09-17 21:28:40] production.INFO: Finance: PIREP: RBJeLydD6ebz4VK8, fare count: {"fare_id":1,"count":152}  
[2024-09-17 21:28:40] production.INFO: Finance: PIREP: RBJeLydD6ebz4VK8, fare count: {"fare_id":2,"count":8941}  
[2024-09-17 21:28:40] production.DEBUG: Looking for DCT  
[2024-09-17 21:28:40] production.DEBUG: Looking for RSK  
[2024-09-17 21:28:40] production.DEBUG: name: RSK - 36.7484x-108.0989  
[2024-09-17 21:28:40] production.DEBUG: Looking for LARKS2  
[2024-09-17 21:28:40] production.INFO: New PIREP filed, pirep_id: RBJeLydD6ebz4VK8  
[2024-09-17 21:28:40] production.INFO: Bid for user: FA003 on flight FA742  
[2024-09-17 21:28:40] production.INFO: NotificationEvents::onPirepFile: RBJeLydD6ebz4VK8 filed  
[2024-09-17 21:28:41] production.INFO: Disposable Basic | Pirep:RBJeLydD6ebz4VK8 FILED, Change STATE of N738FA to PARKED  
[2024-09-17 21:28:41] production.INFO: Disposable Special | Maintenance Records for N738FA (ID:5) increased Pirep ID:RBJeLydD6ebz4VK8 FILED  
```

**Comment by ProAviaAZ** [comment]
Dev build 7.0.38
No emails being sent to admins.
Here's some additional info that may help.

![image](https://github.com/user-attachments/assets/ac6f98e3-12f9-4365-abd2-2dec020aec04)

**Comment by aerocaribbeanva** [comment]
We have v7.0.3 stable release and can confirm we observe the same error on our logs and the pirep email wasn't received by the admins.

I can confirm the pilot is receiving the email notification that the pirep was accepted.

**Comment by nabeelio** [comment]
Above [this line](https://github.com/phpvms/phpvms/blob/main/app/Notifications/Messages/PirepFiled.php#L22), I think calling

```php
$this->pirep->simbrief = null;
```

That should maybe fix it

**Comment by nabeelio** [comment]
If it does, we can do a PR

**Comment by aerocaribbeanva** [comment]
Added the change would let you if the error doesnt show up again in the next days and all pirep admin emails are received.

**Comment by nabeelio** [comment]
You can clear the cache to be safe but that shouldn't be necessary

**Comment by aerocaribbeanva** [comment]
@nabeelio if the pirep instance passed to the notification is `readonly` can we actually modify that field?

![Image](https://github.com/user-attachments/assets/a9786cc7-745e-4023-8faf-869e4cba8c1a)

**Comment by nabeelio** [comment]
You should be able to modify a property, I don't believe this means that the instance is immutable. Did it give an error?

**Comment by aerocaribbeanva** [comment]
No errors, asking because I still saw the same `Error=Serialization of 'App\Models\SimBriefXML' is not allowed`

**Comment by nabeelio** [comment]
Ah okay, I'll take a look at the code and fix it. Thank you for checking that! Guess the easy way won't work

**Comment by aerocaribbeanva** [comment]
Weird thing is that pilot email is receive and the only different thing I see between the two flows is that the blade for the admins has 
```
  A new PIREP has been submitted by {{ $pirep->user->ident }} {{ $pirep->user->name }}
```

but everything else is similar :/

**Comment by ProAviaAZ** [comment]
> No errors, asking because I still saw the same `Error=Serialization of 'App\Models\SimBriefXML' is not allowed` error. Would try clearing the cache

I am seeing the same after adding the code 
$this->pirep->simbrief = null; 
And no admin email received.

**Comment by aerocaribbeanva** [comment]
@ProAviaAZ @nabeelio  I reverted that change because for some reason I also started seeing a `simbrief` column doesn't exist error via the API from the ACARS it showed up at the same time as the email failed 😅

**Comment by nabeelio** [comment]
I'm trying to replicate this in some tests and I can't get it to happen. It's very strange

**Comment by nabeelio** [comment]
Can you try this branch: https://github.com/phpvms/phpvms/pull/2033

**Comment by FatihKoz** [comment]
Maybe using built in laravel features can solve this issue

![Image](https://github.com/user-attachments/assets/0a057b31-de7d-42e0-aafd-9712c536eda8)

https://github.com/phpvms/phpvms/blob/4a7b628e8361182d45a4e296a451f3b797f2d7af/app/Notifications/Messages/PirepFiled.php#L25

Can someone try below change at mentioned line to see if it helps.
```php
['pirep' => $this->pirep->withoutRelations()]
```
Good luck

**Comment by flyteaceo** [comment]
> Maybe using built in laravel features can solve this issue
> 
> ![Image](https://github.com/user-attachments/assets/0a057b31-de7d-42e0-aafd-9712c536eda8)
> 
> [phpvms/app/Notifications/Messages/PirepFiled.php](https://github.com/phpvms/phpvms/blob/4a7b628e8361182d45a4e296a451f3b797f2d7af/app/Notifications/Messages/PirepFiled.php#L25)
> 
> Line 25 in [4a7b628](/phpvms/phpvms/commit/4a7b628e8361182d45a4e296a451f3b797f2d7af)
> 
>  ['pirep' => $this->pirep] 
> Can someone try below change at mentioned line to see if it helps.
> 
> ['pirep' => $this->pirep->withoutRelations()]
> Good luck
> 
> _Reference: https://laravel.com/docs/10.x/queues#handling-relationships_

I tried this and it worked.

**Comment by aerocaribbeanva** [comment]
I'll add it now too 🙏🏾

**Comment by flyteaceo** [comment]
Error came back on second flight and no admin email was sent. Changed nothing between flights.

**Comment by FatihKoz** [comment]
I added three more calls, to where it all starts. Kindly check the change on the PR.

https://github.com/phpvms/phpvms/pull/2036/files

Hoping to nail this down for the last time 😉

**Comment by aerocaribbeanva** [comment]
Would also give it a try later to the patch-2 . I also continued seeing the error with the initial change.
@ProAviaAZ @nabeelio @FatihKoz  only the flights with a simbrief briefing and ofp are the ones affected. The flights where the user doesn't use simbrief are the ones that work fine.

**Comment by FatihKoz** [comment]
@nabeelio the issue may still be there and active.

We got negative (failing) reports even after applying all the changes manually, I believe the serialization error occurs even before we start processing it for the notification.

I had some time to read stuff and found out some PHP limitations/problems about serialization of `SimpleXMLElement` datatype, worked on some quick solutions like casting them as a `string`, that caused more problems on phpVMS side.

`
After PHP 5.3 an object made by SimpleXML_Load_String() cannot be serialized. An attempt to do so will result in a run-time failure, throwing an exception.
`

Rest of the code works fine, we only have problems while loading the pirep for a notification (in which we really do not need the SimBrief relationship at all). And this serialization occurs while loading the SimBrief model itself, which is called by the pirep relationship (yes, even though we say it to not load it for the notification, this happens before it).

Check line 66 of `app\Models\SimBrief.php`

Therefore I would kindly advise keeping this open and digging deeper :)

**Comment by ProAviaAZ** [comment]
@nabeelio 
@FatihKoz 

I agree - this is still an issue after making all changes manually on my install. 
Manual PIREPS notify fine. Automated with SB are the issue.

**Comment by nabeelio** [comment]
Ok, I'll re-open. I haven't been able to replicate it, so it's really quite strange. But I have a different idea.

**Comment by nabeelio** [comment]
In the notificationeventshandler.php, above the `Log::emergency` line, can you add `report($e);`. I want the full exception stack trace. Maybe that will help in where to get the right attribute.

It's being followed/looked up, that's why removing relationships from it isn't helping

![Image](https://github.com/user-attachments/assets/1f45b9a8-2404-43e6-91ec-9343a495412b)

**Comment by ProAviaAZ** [comment]
Done
I’ll get you the log once a PIREP or two comes in

**Comment by nabeelio** [comment]
So the error is actually coming from the bowels of the queue driver I'm guessing. It seems like maybe it's rehydrating. So that column maybe needs to be not selected by default by that relationship, and then explicitly queried (which it might already be?) when it's retrieved for the flight. That's the only workaround I can think of right now

```
[2025-05-06 16:20:06] production.ERROR: Serialization of 'App\Models\SimBriefXML' is not allowed {"userId":9,"exception":"[object] (Exception(code: 0): Serialization of 'App\\Models\\SimBriefXML' is not allowed at /path/vendor/laravel/framework/src/Illuminate/Queue/Queue.php:160)
[stacktrace]
#0 /path/vendor/laravel/framework/src/Illuminate/Queue/Queue.php(160): serialize(Object(Illuminate\\Notifications\\SendQueuedNotifications))
#1 /path/vendor/laravel/framework/src/Illuminate/Queue/Queue.php(129): Illuminate\\Queue\\Queue->createObjectPayload(Object(Illuminate\\Notifications\\SendQueuedNotifications), NULL)
#2 /path/vendor/laravel/framework/src/Illuminate/Queue/Queue.php(107): Illuminate\\Queue\\Queue->createPayloadArray(Object(Illuminate\\Notifications\\SendQueuedNotifications), NULL, '')
#3 /path/vendor/laravel/framework/src/Illuminate/Queue/SyncQueue.php(38): Illuminate\\Queue\\Queue->createPayload(Object(Illuminate\\Notifications\\SendQueuedNotifications), NULL, '')
#4 /path/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(254): Illuminate\\Queue\\SyncQueue->push(Object(Illuminate\\Notifications\\SendQueuedNotifications))
#5 /path/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(230): Illuminate\\Bus\\Dispatcher->pushCommandToQueue(Object(Illuminate\\Queue\\SyncQueue), Object(Illuminate\\Notifications\\SendQueuedNotifications))
#6 /path/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(77): Illuminate\\Bus\\Dispatcher->dispatchToQueue(Object(Illuminate\\Notifications\\SendQueuedNotifications))
#7 /path/vendor/laravel/framework/src/Illuminate/Notifications/NotificationSender.php(229): Illuminate\\Bus\\Dispatcher->dispatch(Object(Illuminate\\Notifications\\SendQueuedNotifications))
#8 /path/vendor/laravel/framework/src/Illuminate/Notifications/NotificationSender.php(76): Illuminate\\Notifications\\NotificationSender->queueNotification(Array, Object(App\\Notifications\\Messages\\PirepFiled))
#9 /path/vendor/laravel/framework/src/Illuminate/Notifications/ChannelManager.php(39): Illuminate\\Notifications\\NotificationSender->send(Array, Object(App\\Notifications\\Messages\\PirepFiled))
#10 /path/vendor/laravel/framework/src/Illuminate/Support/Facades/Facade.php(355): Illuminate\\Notifications\\ChannelManager->send(Array, Object(App\\Notifications\\Messages\\PirepFiled))
#11 /path/app/Notifications/NotificationEventsHandler.php(67): Illuminate\\Support\\Facades\\Facade::__callStatic('send', Array)
#12 /path/app/Notifications/NotificationEventsHandler.php(208): App\\Notifications\\NotificationEventsHandler->notifyAdmins(Object(App\\Notifications\\Messages\\PirepFiled))
#13 /path/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(478): App\\Notifications\\NotificationEventsHandler->onPirepFile(Object(App\\Events\\PirepFiled))
#14 /path/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(286): Illuminate\\Events\\Dispatcher->Illuminate\\Events\\{closure}('App\\\\Events\\\\Pire...', Array)
#15 /path/vendor/laravel/framework/src/Illuminate/Events/Dispatcher.php(266): Illuminate\\Events\\Dispatcher->invokeListeners('App\\\\Events\\\\Pire...', Array, false)
#16 /path/vendor/laravel/framework/src/Illuminate/Foundation/helpers.php(433): Illuminate\\Events\\Dispatcher->dispatch('App\\\\Events\\\\Pire...')
#17 /path/app/Services/PirepService.php(442): event(Object(App\\Events\\PirepFiled))
#18 /path/app/Http/Controllers/Api/PirepController.php(297): App\\Services\\PirepService->submit(Object(App\\Models\\Pirep))
```
